### PR TITLE
Don't reflow the alt buffer on resize

### LINKED ISSUE #3493: Consider not wrapping the alternate screen on resize
# Description of the new feature/enhancement

VTE only rewraps the contents of the (normal screen + its scrollback buffer) on a resize event. It doesn't rewrap the contents of the alternate screen.

Rationale:

The alternate screen is used by applications which repaint it after a resize event. So it doesn't really matter.

However, in that short time window, after resizing the terminal but before the application catches up, this prevents vertical lines (e.g. the user interface of `mc`) from becoming ugly tilted.

### PR DESCRIPTION
**Pull request by zadjii-msft**
VTE only rewraps the contents of the (normal screen + its scrollback
buffer) on a resize event. It doesn't rewrap the contents of the
alternate screen. The alternate screen is used by applications which
repaint it after a resize event. So, it doesn't really matter. However,
in that short time window, after resizing the terminal but before the
application catches up, this prevents vertical lines

It was really hard to get a gif of this where it happened and was small
enough to upload to GH, but there is one in #12719.

There's something in this branch that fixes a scrolling issue in the
parent PR. I'm partially filing this so I can look at the diffs here and
try and figure out what that is. I kinda want to just take all 3 alt
buffer PRs as a single atomic unit, but splitting them up made sense
from a review standpoint.

Closes #3493

### CODE REVIEW

**Comment by zadjii-msft** [comment]
@j4james Okay, so I know I'm looking for your signoff on #12561 and that branch is a bit funky, but does this one seem good to you? I'm not sure where exactly in this PR I fixed the issue in #12561. If this one seems good to you, then I'm just gonna merge the whole stack of alt buffer PR's all at once and call it a day.

File: src/host/screenInfo.cpp [hunk 1]
```diff
@@ -718,9 +718,8 @@ void SCREEN_INFORMATION::SetViewportSize(const COORD* const pcoordSize)
 
         if (_psiMainBuffer)
         {
-            const auto bufferSize = GetBufferSize().Dimensions();
-
-            _psiMainBuffer->SetViewportSize(&bufferSize);
+            _psiMainBuffer->_fAltWindowChanged = false;
+            _psiMainBuffer->_deferredPtyResize = til::size{ GetBufferSize().Dimensions() };
```
Comment (j4james) [inline src/host/screenInfo.cpp:722]: I don't think this is the right place to be doing this. You're deferring the viewport size, but that's not the same thing as the buffer size. I don't know if this is good enough to make the Terminal alt buffer work (I haven't looked at that side), but in cohost the results are a bit bonkers. For example, try something like this:

```
printf "\e[?1049h" # Switch to alt buffer
printf "\e[8;24;60t" # Make the window narrower (assuming 80x24 shrinking to 60x24)
printf "\e[?1049l" # Switch back to the main buffer
```

I'd expect the main buffer to keep that 60x24 size, but what you get is a 80 width buffer in a 60 width window. In other cases it'll just do nothing at all.

And ideally this would be using the same mechanism as the existing `_fAltWindowChanged` deferred sizing, otherwise there are bound to be weird results when you mix the too. I don't know how feasible that is though.

I honestly don't know what the right answer is here, but this definitely feels wrong.

File: src/host/screenInfo.cpp [hunk 2]
```diff
@@ -1981,6 +1990,12 @@ void SCREEN_INFORMATION::UseMainScreenBuffer()
             psiMain->ProcessResizeWindow(&(psiMain->_rcAltSavedClientNew), &(psiMain->_rcAltSavedClientOld));
             psiMain->_fAltWindowChanged = false;
         }
+        else if (_psiMainBuffer->_deferredPtyResize.has_value())
+        {
+            const COORD newSize = _psiMainBuffer->_deferredPtyResize.value().to_win32_coord();
+            _psiMainBuffer->SetViewportSize(&newSize);
+            _psiMainBuffer->_deferredPtyResize = std::nullopt;
+        }
```
Comment (j4james) [inline src/host/screenInfo.cpp:1998]: It's worth noting that the `_fAltWindowChanged` resize is triggered in the `UseAlternateScreenBuffer` as well. I'm not entirely sure why that is, but I would expect this deferred resize to be triggered under the same conditions, unless there's a good reason not to. Of course if we could unify the two somehow then this would be a moot point.

File: src/host/screenInfo.cpp (hunk 1 above)
Comment (zadjii-msft) [reply src/host/screenInfo.cpp:722]: This is exactly the kind of feedback I was looking for. Lemme take a look ☺️

Comment (zadjii-msft) [reply src/host/screenInfo.cpp:722]: (did not find this already on the backlog)

Comment (j4james) [reply src/host/screenInfo.cpp:722]: Yeah, this didn't work in conhost before because `SetViewportSize` wasn't deferring the size requests in alt mode. But that's exactly what you're doing now, so this update should really fix that bug, except that it doesn't. So my concern is that we'd essentially have to rewrite this code to fix it, so why not just do it right the first time?

That said, I don't know what the right solution is, so if it's not obvious to you how to deal with it, it's probably not a big deal to leave it as is. It's obviously not a common use-case else someone would assumedly have filed a bug report by now.

Comment (zadjii-msft) [reply src/host/screenInfo.cpp:722]: You're not wrong 😛 'spose I was trying to selfishly get these bits of the alt buffer in now, and worry about that bug later. I'll give it another go though.

File: src/host/screenInfo.cpp (hunk 2 above)
Comment (zadjii-msft) [reply src/host/screenInfo.cpp]: should be in 764237c15, thanks for the catch

File: src/host/screenInfo.cpp (hunk 1 above)
Comment (zadjii-msft) [reply src/host/screenInfo.cpp:722]: Okay, I think this should be fixed in the latest push now.

**Comment by j4james** [comment]
I've playing around with the latest version and in general it looks good to me, but I came across a weird edge case which triggered a runtime error dialog and then killed the conpty session (at least I think that's what's happening).

Steps to reproduce:
1. Open a bash shell in Windows Terminal.
2. Execute `printf "\e[8;24;132t\e[12;100H\e[?1049h\e[8;10;80t"`

I wasn't actually expecting this to do anything meaningful considering WT can't do resizing, but it happened to be part of one of my alt buffer test scripts, and I wasn't expecting it to crash.

**Comment by zadjii-msft** [comment]
> I came across a weird edge case which triggered a runtime error

well, we definitely shouldn't be able to do that, so I'll dig in ☺️

**Comment by zadjii-msft** [comment]
Okay, that crash is fixed now too, thanks again @lhecker for the assist ☺️

File: src/cascadia/TerminalCore/Terminal.cpp [hunk 3]
```diff
@@ -237,12 +237,38 @@ void Terminal::UpdateAppearance(const ICoreAppearance& appearance)
...
+    if (_inAltBuffer())
+    {
+        // stash this resize for the future.
+        _deferredResize = til::size{ viewportSize };
+
+        _altBuffer->GetCursor().StartDeferDrawing();
+        // we're capturing `this` here because when we exit, we want to EndDefer on the (newly created) active buffer.
+        auto endDefer = wil::scope_exit([this]() noexcept { _altBuffer->GetCursor().EndDeferDrawing(); });
+
+        // GH#3494: We don't need to reflow the alt buffer. Apps that use the
```
Comment (miniksa) [inline src/cascadia/TerminalCore/Terminal.cpp:259]: Beautiful.

File: src/cascadia/TerminalCore/Terminal.hpp [hunk 4]
```diff
@@ -328,6 +328,9 @@ class Microsoft::Terminal::Core::Terminal final :
     Microsoft::Console::Types::Viewport _mutableViewport;
     SHORT _scrollbackLines;
 
+    til::size _altBufferSize;
```
Comment (miniksa) [inline src/cascadia/TerminalCore/Terminal.hpp:331]: Yeah that makes sense. In hindsight, it's obvious that it was easier to mess this up when both buffers were sharing.

File: src/cascadia/UnitTests_TerminalCore/ConptyRoundtripTests.cpp [hunk 5]
```diff
@@ -4068,3 +4070,47 @@ void ConptyRoundtripTests::AltBufferToAltBufferTest()
...
+    auto& sm = si.GetStateMachine();
+
+    gci.LockConsole(); // Lock must be taken to manipulate alt/main buffer state.
+    auto unlock = wil::scope_exit([&] { gci.UnlockConsole(); });
+
+    _flushFirstFrame();
+
+    _checkConptyOutput = false;
+
+    // Note: we really don't care about the output in this test. We could, but
```
Comment (miniksa) [inline src/cascadia/UnitTests_TerminalCore/ConptyRoundtripTests.cpp:4093]: `VERIFY_DID_NOT_CRASH()` lol.

**Review by miniksa — APPROVED** [review]
Looks good!

**Comment by DHowett** [comment]
@msftbot make sure @dhowett signs off

**Comment by DHowett** [comment]
I pushed a difficult merge, but I verified that the only change between the original diff base and the new diff base is the new commit added in the alt buffer branch

**Review by lhecker — APPROVED** [review]

**Comment by DHowett** [comment]
Above: conhost inbox. Notice that as I go wild on the resize, there's a
few frames where the text is all _slanty_, as the buffer is reflowed,
before `mc` has a chance to redraw.  Below: this PR. no more _slanties_. 

gh-3493-conhost

(The effect is much more noticable IRL).

File: src/cascadia/TerminalCore/Terminal.cpp [hunk 6]
```diff
@@ -972,9 +959,14 @@ int Terminal::_VisibleEndIndex() const noexcept
 
 Viewport Terminal::_GetVisibleViewport() const noexcept
 {
+    // GH#3493: if we're in the alt buffer, then it's possible that the mutable
+    // viewport's size hasn't been updated yet. In that case, use the
+    // temporarily stashed _altBufferSize instead.
     const COORD origin{ 0, gsl::narrow<short>(_VisibleStartIndex()) };
+    const COORD size{ _inAltBuffer() ? _altBufferSize.to_win32_coord() :
```
Comment (DHowett) [inline src/cascadia/TerminalCore/Terminal.cpp:966]: This is almost aching for another helper ... or, seems like you could use the Dimensions off the GetMutableViewport helper's return value? I'd prefer concentrating the magic in one place. :)

File: src/host/screenInfo.cpp [hunk 7]
```diff
@@ -1190,20 +1189,6 @@ void SCREEN_INFORMATION::_InternalSetViewportSize(const COORD* const pcoordSize,
                 // Adjust the top of the window instead of the bottom
                 // (so the lines slide upward)
                 srNewViewport.Top -= DeltaY;
-
-                // If we happened to move the top of the window past
-                // the 0th row (first row in the buffer)
-                if (srNewViewport.Top < 0)
```
Comment (DHowett) [inline src/host/screenInfo.cpp:1196]: Don't totally understand why this disappeared

**Review by DHowett — CHANGES_REQUESTED** [review]
I'm inclined to trust the resizing code in conhost here, even though it scares me. I don't love that we have to reimplement part of ResizeWindow...

However, there's two concerning uses of `_mutableViewport` after you've fully decoupled them:

mouse event clamping
https://github.com/microsoft/terminal/blob/098f6e305b0742a8a25b46e8f669fed726b26d21/src/cascadia/TerminalCore/Terminal.cpp#L696

this whole block in AdjustCursorPosition
https://github.com/microsoft/terminal/blob/098f6e305b0742a8a25b46e8f669fed726b26d21/src/cascadia/TerminalCore/Terminal.cpp#L1117-L1128

Are those worth worrying about? I can imagine a world where we make the window wider in the alt buffer but mouse events are clamped to the smaller main buffer and fail to work properly after resize.

(Do we have tests covering this?)

File: src/cascadia/TerminalCore/Terminal.cpp (hunk 6 above)
Comment (lhecker) [reply src/cascadia/TerminalCore/Terminal.cpp:966]: Can't we make a `til::rect _viewports[2]` and index into it with a `bool _altBufferInUse`?
Like... `_viewports[_altBufferInUse].to_win32_coord()` or something like that...

File: src/host/screenInfo.cpp (hunk 7 above)
Comment (lhecker) [reply src/host/screenInfo.cpp:1196]: I removed that code while debugging the crash issue.
Check out my commit here: [`4c1404d` (#12719)](https://github.com/microsoft/terminal/pull/12719/commits/4c1404d524dd1ba437bc681bd3b3095b5fe38f95)

Line 1227-1231 it says:
```cpp
if (srNewViewport.Top < 0)
{
    srNewViewport.Bottom -= srNewViewport.Top;
    srNewViewport.Top = 0;
}
```

That if condition makes the removed code with its fail-fast, etc. redundant.

File: src/cascadia/TerminalCore/Terminal.cpp (hunk 6 above)
Comment (j4james) [reply src/cascadia/TerminalCore/Terminal.cpp:966]: Bear in mind that we'll probably need to support more buffers than just main and alt in the long term. So I'd expect properties like this would ultimately be part of the text buffer itself, and then you could just do something like `_activeBuffer().mutableViewport`. But we also need to get the architecture synced up with conhost, which currently manages this sort of state in the `SCREEN_INFORMATION` class. Not that I'm saying this is essential for now, but if you're trying to clean up the code, it's best to also consider how we are likely to extend this in the future.

File: src/host/screenInfo.cpp (hunk 7 above)
Comment (DHowett) [reply src/host/screenInfo.cpp:1196]: I didn't look at that block you introduced below as being related since the comment said "off the right" instead of "off the bottom"!

File: src/cascadia/UnitTests_Control/ControlInteractivityTests.cpp [hunk 8]
```diff
@@ -846,11 +852,99 @@ namespace ControlUnitTests
...
+        Log::Comment(L"Resize the terminal to be 10 columns wider");
+        core->SizeChanged(40, 20);
+
+        Log::Comment(L"Click on a spot that's NOW INSIDE the buffer");
+        expectedOutput.push_back(L"\x1b[M C&");
+        interactivity->PointerPressed(leftMouseDown,
+                                      WM_LBUTTONDOWN, //pointerUpdateKind
+                                      0, // timestamp
+                                      modifiers,
+                                      cursorPosition1.to_core_point());
```
Comment (DHowett) [inline src/cascadia/UnitTests_Control/ControlInteractivityTests.cpp:966]: Callback-based tests scare me a little bit, because it is so easy to author one that *fires no input* (due to the bug under test) and the callback never fires to validate it. That's why I added "verifyExpectedInputDrained" to one of the test fixtures...
Do we need a guarantee like that here? How do we know that PointerPressed emitted anything?

File: src/cascadia/UnitTests_Control/ControlInteractivityTests.cpp [hunk 9]
```diff
@@ -107,6 +109,10 @@ namespace ControlUnitTests
                 Log::Comment(fmt::format(L"Expected: \"{}\"", TerminalCoreUnitTests::TestUtils::ReplaceEscapes(expected)).c_str());
                 VERIFY_ARE_EQUAL(expected, hstr);
             });
+
+            return std::move(wil::scope_exit([&]() {
+                VERIFY_ARE_EQUAL(0, expectedOutput.size(), L"Validate we drained all the expected output");
```
Comment (DHowett) [inline src/cascadia/UnitTests_Control/ControlInteractivityTests.cpp:114]: NEVERMIND LOL YOU ADDED IT

File: src/cascadia/UnitTests_Control/ControlInteractivityTests.cpp [hunk 10]
```diff
@@ -924,27 +919,48 @@ namespace ControlUnitTests
...
+                                      modifiers,
+                                      cursorPosition0.to_core_point());
+        // Flush it out.
+        conn->WriteInput(L"sentinel");
+        VERIFY_ARE_EQUAL(0, expectedOutput.size(), L"Validate we drained all the expected output");
 
-        Log::Comment(L"Switch to alt buffer");
+        Log::Comment(L" --- Switch to alt buffer ---");
         term.Write(L"\x1b[?1049h");
+        auto returnToMain = wil::scope_exit([&]() { term.Write(L"\x1b[?1049h"); });
```
Comment (DHowett) [inline src/cascadia/UnitTests_Control/ControlInteractivityTests.cpp:944]: we could just give these all Method-level isolation..!

**Review by DHowett — APPROVED** [review]
Spellbot hates it, but I think I love it.

**Comment by DHowett** [comment]
More importantly, it doesn't build on x86 :D

File: src/cascadia/UnitTests_Control/ControlInteractivityTests.cpp (hunk 10 above)
Comment (zadjii-msft) [reply src/cascadia/UnitTests_Control/ControlInteractivityTests.cpp:944]: after watching VS choke on going in and out of debugging on method-level isolation all day, _nah this is close enough_

File: src/cascadia/UnitTests_Control/ControlInteractivityTests.cpp (hunk 9 above)
Comment (zadjii-msft) [reply src/cascadia/UnitTests_Control/ControlInteractivityTests.cpp:114]: 🤣

**Review by DHowett — APPROVED** [review]

**Comment by DHowett** [comment]
Did we break CI with the Code Thing PR?
